curve25519-dalek/src/elligator2.rs

```diff
@@ -0,0 +1,786 @@
+// us to match the existing [`PublicKey`] generation interface for the
+// [`PublicRepresentative`] in the [`x25519_dalek`] crate. AFAIK there is no
+// need for anyone with only the public key to be able to generate the
+// representative.
+pub(crate) fn v_in_sqrt(key_input: &[u8; 32]) -> Choice {
+    let mut masked_pk = *key_input;
+    masked_pk[0] &= 0xf8;
+    masked_pk[31] &= 0x7f;
+    masked_pk[31] |= 0x40;
+
+    let pubkey = EdwardsPoint::mul_base_clamped(masked_pk);
+    v_in_sqrt_pubkey_edwards(&pubkey)
+}
+
+/// Determines if `V <= (p - 1)/2` for an EdwardsPoint (e.g an x25519 public key)
+/// and returns a [`Choice`] indicating the result.
+pub(crate) fn v_in_sqrt_pubkey_edwards(pubkey: &EdwardsPoint) -> Choice {
+    let divide_minus_p_1_2 = FieldElement::from_bytes(&DIVIDE_MINUS_P_1_2_BYTES);
+
+    // sqrtMinusAPlus2 is sqrt(-(486662+2))
+    let (_, sqrt_minus_a_plus_2) = FieldElement::sqrt_ratio_i(
+        &(&MONTGOMERY_A_NEG - &(&FieldElement::ONE + &FieldElement::ONE)),
+        &FieldElement::ONE,
+    );
+
+    // inv1 = 1/((A.Z - A.y) * A.X)
+    let inv1 = (&(&pubkey.Z - &pubkey.Y) * &pubkey.X).invert();
+
+    // t0 = A.Y + A.Z
+    let t0 = &pubkey.Y + &pubkey.Z;
+
+    // v = t0 * inv1 * A.Z * sqrtMinusAPlus2
+    let v = &(&t0 * &inv1) * &(&pubkey.Z * &sqrt_minus_a_plus_2);
+
+    // is   v <= (q-1)/2  ?
+    divide_minus_p_1_2.ct_gt(&v)
+}
+
+// ============================================================================
+// ============================================================================
+
+fn map_to_curve_parts(
+    r: &FieldElement,
+) -> (FieldElement, FieldElement, FieldElement, FieldElement) {
+    let zero = FieldElement::ZERO;
+    let one = FieldElement::ONE;
+    let minus_one = -&FieldElement::ONE;
+
+    // Exceptional case 2u^2 == -1
+    let mut tv1 = r.square2();
+    tv1.conditional_assign(&zero, tv1.ct_eq(&minus_one));
+
+    let d_1 = &one + &tv1; /* 1 + 2u^2 */
+    let d = &MONTGOMERY_A_NEG * &(d_1.invert()); /* d = -A/(1+2u^2) */
+
+    let inner = &(&d.square() + &(&d * &MONTGOMERY_A)) + &one;
+    let gx1 = &d * &inner; /* gx1 = d^3 + Ad^2 + d */
+    let gx2 = &gx1 * &tv1;
+
+    let eps_is_sq = high_y(&d);
+
+    // complete X
+    /* A_temp = 0, or A if nonsquare*/
+    let a_temp = FieldElement::conditional_select(&MONTGOMERY_A, &zero, eps_is_sq);
+    let mut x = &d + &a_temp; /* d, or d+A if nonsquare */
+    x.conditional_negate(!eps_is_sq); /* d, or -d-A if nonsquare */
+
+    // complete Y
+    let y2 = FieldElement::conditional_select(&gx2, &gx1, eps_is_sq);
+    let (_, mut y) = FieldElement::sqrt_ratio_i(&y2, &one);
+    y.conditional_negate(eps_is_sq ^ y.is_negative());
+
+    (&x * &d_1, d_1, y, one)
+}
+
+pub(crate) fn map_fe_to_montgomery(r: &FieldElement) -> (FieldElement, FieldElement) {
+    let (xmn, xmd, y, _) = map_to_curve_parts(r);
+    (&xmn * &(xmd.invert()), y)
+}
+
+pub(crate) fn map_fe_to_edwards(r: &FieldElement) -> (FieldElement, FieldElement) {
```

Comment:
Should mention this is `map_to_curve_elligator2_edwards25519` from RFC 9380 section G.2.2